_docs/software_support_lifecycle.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: Software Support Lifecycle
+date: 2020-12-09
+summary: Vox Pupuli policy on supporting upstream software
+---
+
+## Software Support Lifecycle
+
+As a part of Vox Pupuli's mission to provide a home for modules that still have a userbase and need development, we need to define a clear support strategy for what operating systems, dependencies, and upstream software we will support and for how long those will be supported.
+
+As we operate entirely on the graces of our community members being willing to voluntarily contribute to our modules, we also need to be respective of their time and resources and not add undue stress in order to support software that upstream creators no longer support.
+
+To account for this, we no longer support any software or operating system after the date that the maintainers have marked said software or operating system as End of Life. However, this does not mean we will immediately remove code that allows for the module to run on End of Life'd software or operating systems, but the community will no longer be expected to provide support to users for that software.
+
+There are exceptions and the full policy is detailed in the next section.
+
+### Support Lifecycle
+
+#### Operating Systems
+An Operating System is no longer considered supported once it reaches an End of Life date determined by the upstream maintainer* of the operating system. To ensure our users have enough time to adjust, we will take the following steps to ensure as seamless of a transition as possible:
+
+1. A deprecation PR will be created no later that 3 months before an Operating System's End of Life date.
```

Comment:
Do you mean "no earlier than" ?

Reply:
I think a PR 3 months in advance is good. However, is the intent also to merge it or is it to open a discussion? For a discussion, an issue may be good.

Perhaps this is also something we can script, using metadata.json. They should also have a label so it's easy to find them.

Reply:
I said "No later than", so that ample time is given to the user. We had a long discussion on the PMC call about it recently and the idea was that, like any Enterprise software, if a user wants support post-EoL date, they have to pay a vendor or support it themselves.

Vox Pupuli doesn't have the manpower to support EoL OSes and Software. It'd be different if every EoL date was like CentOS 8, but in reality, most of those dates are out there a year or more ahead of time, so I'm strongly against supporting past an EOL date.

However, "End of Support" does not mean we should break it intentionally either. We leave functionality in there, if possible, until 1 year after the EOL date. The exception being that if a currently supported/stable version needs something that breaks the EoL version, then the supported versions take precedence.

Reply:
I'm opposed to having dead code in modules, but it's not black and white. I'll give some examples that I can think of. In the following examples OS major version 6 is EOL, 7 isn't.

```puppet
# No need to change 6 to 7, that's not really a difference and only creates noise in the git log
if versioncmp($facts['os']['release']['major'], '6') > 0 {}

# This is code that's not run on any supported OS which means it's not covered by tests
if $facts['os']['release']['major'] == '6' {}
```

So while it's annoying, I do think the rule of thumb should be that we can test all code.

I also think that individual contributors should be able to support it if they provide the manpower to keep it supported. Let's say I need a module on an EOL OS, then if I make sure it's all working, that should be OK. This should be documented somewhere (README?). It is also an explicit process. During the period before it goes EOL, it must be noted by a project member that they are interested in keeping it up. Given our current project composition, this may not be used that much. I know I generally have no interest in it since I keep my distros supported.

Reply:
I would also add the caveat that keeping up code to work with an EOL OS would only be allowed as long as it doesn't prevent the project from moving forward to add features or support current OSes.

Reply:
Btw, I think https://github.com/puppetlabs/puppetlabs-apache is a good example of the cruft that can build up. It still has conditionals for Ubuntu 10.04.